lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystemCache.java

```diff
@@ -0,0 +1,218 @@
+
+public final class MemoryFileSystemCache
+        implements TrinoFileSystemCache
+{
+    private final Tracer tracer;
+    private final Cache<String, Optional<Slice>> cache;
+    private final int maxContentLengthBytes;
+    private final AtomicLong largeFileSkippedCount = new AtomicLong();
+
+    @Inject
+    public MemoryFileSystemCache(Tracer tracer, MemoryFileSystemCacheConfig config)
+    {
+        this(tracer, config.getCacheTtl(), config.getMaxSize(), config.getMaxContentLength());
+    }
+
+    private MemoryFileSystemCache(Tracer tracer, Duration expireAfterWrite, DataSize maxSize, DataSize maxContentLength)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+        checkArgument(maxContentLength.compareTo(DataSize.of(1, GIGABYTE)) <= 0, "maxContentLength must be less than or equal to 1GB");
+        this.cache = EvictableCacheBuilder.newBuilder()
+                .maximumWeight(maxSize.toBytes())
+                .weigher((Weigher<String, Optional<Slice>>) (key, value) -> toIntExact(estimatedSizeOf(key) + sizeOf(value, Slice::getRetainedSize)))
+                .expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS)
+                .shareNothingWhenDisabled()
+                .recordStats()
+                .build();
+        this.maxContentLengthBytes = toIntExact(maxContentLength.toBytes());
+    }
+
+    @Override
+    public TrinoInput cacheInput(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getOrLoadFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateInput");
+            return withTracing(span, delegate::newInput);
+        }
+
+        return new MemoryInput(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public TrinoInputStream cacheStream(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getOrLoadFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateStream");
+            return withTracing(span, delegate::newStream);
+        }
+
+        return new MemoryInputStream(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public long cacheLength(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getOrLoadFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateLength");
+            return withTracing(span, delegate::length);
+        }
+
+        return cachedEntry.get().length();
+    }
+
+    @Override
+    public void expire(Location location)
```

Comment:
is there a test for it?

Reply:
There are many existing tests which fail if this method fails to remove entries as expected

Reply:
Could the tests be rectified?

Reply:
The tests are not wrong, they expect normal filesystem behaviour for things like access after delete/update. We would need to disable memory cache for all such tests. It's cleaner to implement expiry and have normal behaviour with memory cache.